### DISCUSSION TITLE
ROS Airsim navigation stack?

### DISCUSSION THREAD

**Discussion by Hi-Peng** [microsoft/AirSim#3275] [Support Q&A] [answered]
Hello everyone, How to use navigation stack on Airsim ROS? thank you

**Comment by GimpelZhang** [comment, marked as answer]
(If you are talking about the car SimMode, which means 2d navigation) Just some suggestions: In my opinion, first you need a map file of the environment you use to simulate. 

Since it is not easy now to get a pgm groundtruth map from the AirSim simulation environment, maybe you have to establish a map first by some slam methods. 

Then you can feed the map to the ros map server and launch the navigation stack. And don't forget to interpret the velocity command from the planner into the "CarControls" message type.

**Reply by jonyMarino** [reply]
I think it could be possible with the new [voxel grid construction](https://github.com/microsoft/AirSim/pull/3239). Please take a look and tell us if it is suitable for this.

**Reply by GimpelZhang** [reply]
> I think it could be possible with the new [voxel grid construction](https://github.com/microsoft/AirSim/pull/3239). Please take a look and tell us if it is suitable for this.

Thanks! I believe that is quite a useful tool.

**Reply by p00uya** [reply]
hi 
I tried all way to run this script on python : 
```
import airsim
c = airsim.VehicleClient()
center = airsim.Vector3r(0, 0, 0)
output_path = os.path.join(os.getcwd(), "map.binvox")
c.simCreateVoxelGrid(center, 100, 100, 100, 0.5, output_path)
```
all steps  are done, but i have an error that said :
 `AttributeError: 'VehicleClient' object has no attribute 'simCreateVoxelGrid'`
@jonyMarino 
how can solve that?

**Reply by GimpelZhang** [reply]
> hi
> I tried all way to run this script on python :
> 
> ```
> import airsim
> c = airsim.VehicleClient()
> center = airsim.Vector3r(0, 0, 0)
> output_path = os.path.join(os.getcwd(), "map.binvox")
> c.simCreateVoxelGrid(center, 100, 100, 100, 0.5, output_path)
> ```
> 
> all steps are done, but i have an error that said :
> `AttributeError: 'VehicleClient' object has no attribute 'simCreateVoxelGrid'`
> @jonyMarino
> how can solve that?

Maybe you can try:

`import setup_path`

before

`import airsim`

**Reply by razekmaiden** [reply]
Hi @p00uya , I had a similar problem. To solve this, I installed the airsim (python) module that is inside of "AirSim/PythonClient" folder instead of the airsim module installed with pip.

You must use the "setup.py" inside of the "AirSim/PythonClient" folder and follow the instructions. I hope you can solve your problem

**Reply by jonyMarino** [reply]
Thanks to all who are carrying out this discussion, it is very enriching for the community!
The pip package was updated. Please try with
 `pip install airsim --upgrade`

**Comment by razekmaiden** [comment]
Hello

I am also looking for a way to integrate ROS navstack to the simulated vehicle in AirSim. My experience is not much, I have been learning and testing for a couple of weeks without good results. I seek to obtain something similar to the typical ROS example of giving objectives to a differential robot in Rviz and that it finds its route.

Excuse me @Hi-Peng , have you had any kind of breakthrough or any advice that you can share?. I would appreciate any information

Another query. Is voxel grid a tool that allows you to generate the map or something like that?

Thank you

**Reply by leoborgnino** [reply]
Hello, I'm working on AirSim+ROS developing MonoSLAM and PathPlanning algorithms. I'm using Multirotor simulation but I think some implementations could be useful for you.
1. In path planning I get the binvox through API, this binvox (answering your question) is a voxel grid (3d map) compressed of your environment. I convert that binvox to an octomap and then I get an array of 3d points of the obstacles.
2. We use our customs path planning algorithms with that 3d map of the environment as input (navigation stack AFAIK only works for 2D navigation) .

Our implementation: 
![planning_drone](https://user-images.githubusercontent.com/8973032/115608349-868ef800-a2bc-11eb-9976-0273a4b98b60.png)

Navigation stack:
![imagen](https://user-images.githubusercontent.com/8973032/115610817-c0153280-a2bf-11eb-8c51-2afe7d5df64b.png)

I don't know if you have installed navstack package in your ROS environment (https://wiki.ros.org/navigation/Tutorials/RobotSetup). After that, you should use the information from airsim sensor topics to connect them to navstack (probably you should do some transformation to match msgs type). You should remove PID control, its cmd vel output is connected to a topic, which calls to AirSim API in order to move the car. Then, connect navstack cmd vel output to that topic. I think you don't need the map to begin the navigation. 

I don't know if this could help you. Could you tell us what have you done? If you have a commit of your changes in ROS I could take a look.

**Reply by razekmaiden** [reply]
Thanks for the answer, I will try to replicate the steps you mention. I already have a guide to continue working.

Greetings!

**Reply by razekmaiden** [reply]
Hello. I have a problem creating the "voxel grid". Basically I have the following message after execute the code described in this page: [Voxel Grid Generator](https://microsoft.github.io/AirSim/voxel_grid/).

**Code:**
```
import airsim
c = airsim.VehicleClient()
center = airsim.Vector3r(0, 0, 0)
output_path = os.path.join(os.getcwd(), "map.binvox")
c.simCreateVoxelGrid(center, 100, 100, 100, 0.5, output_path)
```

**Traceback:**
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'VehicleClient' object has no attribute 'simCreateVoxelGrid'
```


I tried upgrading Unreal from (4.24.3 to 4.25.4), also I recompile AirSim (1.4.0) and reinstall the python dependencies in a virtualEnv, but the problem persist. Is there an specific version to find the createVoxelGrid(), function?

PS: I tried with simCreateVoxelGrid, createVoxelGrid, CreateVoxelGrid

**Reply by leoborgnino** [reply]
It should work (my python code looks exactly like the one you posted), I had to pull and rebuild the last version of airsim APIs in order to get the simCreateVoxelGrid API working. You can see if the repo has the API at AirSim/AirLib/src/api/RpcLibClientBase.cpp. 

@jonyMarino do you know a way to see airsim version from python? the airsim.__version__ doesn't work

**Reply by razekmaiden** [reply]
I already discovered the problem, I was using the AirSim python module installed directly with pip and not the version that came with the repository. For this reason the AirSim module was not properly updated.

**Reply by razekmaiden** [reply]
I have a doubt, the tool "binvox2bt" to convert the map.binvox to map.bt. Is it the one installed with the ROS "octomap" package?

I am having some problems with this conversion ([see](https://github.com/OctoMap/octomap/issues/332)), apparently it returns an empty map.bt.  Am I missing any argument? Is it necessary to download the tool from elsewhere?

**Reply by jonyMarino** [reply]
> It should work (my python code looks exactly like the one you posted), I had to pull and rebuild the last version of airsim APIs in order to get the simCreateVoxelGrid API working. You can see if the repo has the API at AirSim/AirLib/src/api/RpcLibClientBase.cpp.
> 
> @jonyMarino do you know a way to see airsim version from python? the airsim.**version** doesn't work

Sorry for the late response. I'm glad you were able to run the voxel grid generator. We should add the version attribute but it's going to start from that moment.
To disambiguate which airsim version you are using run this command in the script folder
```
pip show airsim
```

**Reply by Leon-Redfield** [reply]
> Hello, I'm working on AirSim+ROS developing MonoSLAM and PathPlanning algorithms. I'm using Multirotor simulation but I think some implementations could be useful for you.
> 
>     1. In path planning I get the binvox through API, this binvox (answering your question) is a voxel grid (3d map) compressed of your environment. I convert that binvox to an octomap and then I get an array of 3d points of the obstacles.
> 
>     2. We use our customs path planning algorithms with that 3d map of the environment as input (navigation stack AFAIK only works for 2D navigation) .
> 
> 
> Our implementation:
> ![planning_drone](https://user-images.githubusercontent.com/8973032/115608349-868ef800-a2bc-11eb-9976-0273a4b98b60.png)
> 
> Navigation stack:
> ![imagen](https://user-images.githubusercontent.com/8973032/115610817-c0153280-a2bf-11eb-8c51-2afe7d5df64b.png)
> 
> I don't know if you have installed navstack package in your ROS environment (https://wiki.ros.org/navigation/Tutorials/RobotSetup). After that, you should use the information from airsim sensor topics to connect them to navstack (probably you should do some transformation to match msgs type). You should remove PID control, its cmd vel output is connected to a topic, which calls to AirSim API in order to move the car. Then, connect navstack cmd vel output to that topic. I think you don't need the map to begin the navigation.
> 
> I don't know if this could help you. Could you tell us what have you done? If you have a commit of your changes in ROS I could take a look.

Hello! I am also connecting ROS with Airsim and applying navigation algorithms. Recently I am having problems getting the global map in PointCloud2. I read the code about "simcreateVoxelGird" but I found it is only useful for very regular objects such as cubes, while for irregular objects it doesn't work well, so I wonder if you have tried some other ways to get the ground truth of the map, thank you!

**Reply by leoborgnino** [reply]
Hi! Did you try reducing "res" parameter in simCreateVoxelGrid(self, position, x, y, z, res, of) ? I didn't try it with small resolution of voxels, so maybe you have tried this and the results are also wrong.

**Reply by razekmaiden** [reply]
Hello everybody. I had to step away from the project for a moment, which gave me time to study ROS a bit more.

Once the "binvox.bt" map was created, I was able to load it into rviz using octomap server. Additionally, with the lidar data (sensor_msgs/PointCloud2) it was possible to continue reconstructing the map. Here is a [video](https://youtu.be/TGMFY2j1tRo).

My next step has been to try to configure the "amcl" node, however when performing the first execution tests, it returns the following errors.

```
[ WARN] [1622124640.277734123, 1621550756.009623889]: Request for map failed; trying again...
[ WARN] [1622124640.778446816, 1621550756.510248287]: Request for map failed; trying again...
[ WARN] [1622124641.285029594, 1621550757.016516024]: Request for map failed; trying again...
[ WARN] [1622124641.787714925, 1621550757.519941715]: Request for map failed; trying again...
[ WARN] [1622124642.291005310, 1621550758.023149508]: Request for map failed; trying again...
[ WARN] [1622124642.794050489, 1621550758.526352309]: Request for map failed; trying again...
```

Is this because the "amcl" node expects the "map_server" node to be running? So far I have only used the "octomap_server" node.  In that case, will it be necessary to export the map generated by octomap_server with "map_server map_saver" to later execute it with the node "map_server"?

A map similar to what is shown below (I do not know why it considers the center as an occupied area):
![image](https://user-images.githubusercontent.com/7081707/119842293-8d6ce400-bed4-11eb-8ddf-0262231af9e4.png)

The goal I have in mind is to configure the [teb local planner](http://wiki.ros.org/teb_local_planner) , which supports robots with ackerman geometry.

I appreciate any opinion / guidance regarding what I am trying to do.
I currently have the repository a bit messy, but if anyone wants to see the roslaunch, or rosbag, or the map file, just ask for it.

PS: I took these courses on udemy:  [Course 1](https://www.udemy.com/course/ros-essentials/), [Course 2](https://www.udemy.com/course/ros-navigation/). For someone who is learning they are highly recommended, although like most of the courses and material available the examples only cover differential robots.

**Reply by razekmaiden** [reply]
Hello, I share some small advances that I have had. As I mentioned previously, I am trying to use the "local teb planner" with the simulated car in AirSim. I am relying on the example files from teb local planner for like car robot. I have modified some parameters based on the approximate dimensions of the simulated vehicle.

Here is the [repository](https://github.com/razekmaiden/AirSim.git) , in case someone wants to see (ros/src/tracer_planner). I also attach a [video](https://youtu.be/msIAcTuiWVw). As you can see, the vehicle can be oriented, but the translation is in small jumps.

I think I am on the right track. I hope that the vehicle control can be solved by better adjusting the planner parameters.

**Reply by alonfaraj** [reply]
@razekmaiden thanks for sharing  your work!
Did you manage to make any progress regrading the small jumps/oscillations?

**Reply by razekmaiden** [reply]
Hi @alonfaraj.  Basically, I perform a lot of test changing some parameters. Until now, I think there is a little improvement in speed and stability in foreward displacement. However when the vehicle requires move to backward, it seems that cannot accomplish the local planner goals, and continues moving until crash with and obstacle.

Some parameters with what I have tried are:
- weight_max_vel_x
- yaw_goal_tolerance
- xy_goal_tolerance
- controller_frequency

**Reply by alonfaraj** [reply]
Thanks!
I will give it a shot.
Do you mind to share your lidar settings from `settings.json`?

**Reply by razekmaiden** [reply]
Hi all.

I started to review each of the nodes and soon I realized that apparently I do not have the amcl node properly configured since it does not publish data. This apparently is related to the conversion from NED (airsim) to ENU (ROS).
If for example I run the roslaunch "airsim_node.launch" by default, I can see that the X axis coincides with the axis of displacement of the vehicle (auto), however when executing the roslaunch "airsim_node.launch" with the parameter "/ airsim_node / odom_frame_id "configured as" odom_local_enu ", the X axis is to one side of the vehicle. This has me confused. If I visualize odometry in rviz, I see something like that.

**with: odom_local_ned**
![image](https://user-images.githubusercontent.com/7081707/127682449-aee8c2b8-257d-48f0-a60b-caa7b8d389df.png)


**with: odom_local_enu**
![image](https://user-images.githubusercontent.com/7081707/127682490-4daf609f-1bff-4a77-a8fe-8c4f013c7213.png)


Have any of you managed to configure this node successfully?
Any comment is useful.

**Comment by lknnnli** [comment]
Hello, I think that's because your old airsim version, try to pip uninstall airsim and then pip install airsim, your problem would be solved.
